src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/NamingStyle.cs

```diff
@@ -0,0 +1,289 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Roslyn.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.Serialization;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles
+{
+    [DataContract]
+    internal class NamingStyle : INamingStyle
+    {
+        [DataMember]
+        public Guid ID { get; set; }
+        [DataMember]
+        public string Name { get; set; }
+        [DataMember]
+        public string Prefix { get; set; }
+        [DataMember]
+        public string Suffix { get; set; }
+        [DataMember]
+        public string WordSeparator { get; set; }
+        [DataMember]
+        public Capitalization CapitalizationScheme { get; set; }
+
+        public IEnumerable<string> InferWords(string name)
+        {
+            // TODO
+            return SpecializedCollections.SingletonEnumerable(name);
+        }
+
+        public string CreateName(IEnumerable<string> words)
+        {
+            EnsureNonNullProperties();
+
+            var wordsWithCasing = ApplyCapitalization(words);
+            var combinedWordsWithCasing = string.Join(WordSeparator, wordsWithCasing);
+            return Prefix + combinedWordsWithCasing + Suffix;
+        }
+
+        private IEnumerable<string> ApplyCapitalization(IEnumerable<string> words)
+        {
+            switch (CapitalizationScheme)
+            {
+                case Capitalization.PascalCase:
+                    return words.Select(w => CapitalizeFirstLetter(w));
+                case Capitalization.CamelCase:
+                    return words.Take(1).Select(w => DecapitalizeFirstLetter(w)).Concat(words.Skip(1).Select(w => CapitalizeFirstLetter(w)));
+                case Capitalization.FirstUpper:
+                    return words.Take(1).Select(w => CapitalizeFirstLetter(w)).Concat(words.Skip(1).Select(w => DecapitalizeFirstLetter(w)));
+                case Capitalization.AllUpper:
+                    return words.Select(w => w.ToUpper());
+                case Capitalization.AllLower:
+                    return words.Select(w => w.ToLower());
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private string CapitalizeFirstLetter(string word)
+        {
+            var chars = word.ToCharArray();
+            return new string(chars.Take(1).Select(c => char.ToUpper(c)).Concat(chars.Skip(1)).ToArray());
+        }
+
+        private string DecapitalizeFirstLetter(string word)
+        {
+            var chars = word.ToCharArray();
+            return new string(chars.Take(1).Select(c => char.ToLower(c)).Concat(chars.Skip(1)).ToArray());
+        }
+
+        public bool IsNameCompliant(string name, out string failureReason)
```

Comment:
i'd rename this to localizedFailureReason.  Or, have an actual enum for this, and later map the enum to the loc string for display purposes.